toolkit/tools/internal/retry/retry.go

```diff
@@ -4,13 +4,39 @@
 package retry
 
 import (
+	"math"
 	"time"
 )
 
+// calculateDelay calculates the delay for the given failure count, sleep duration, and backoff exponent base.
+// If the base is negative, it will calculate a linear backoff.
+// If the base is positive, it will calculate an exponential backoff.
+func calculateDelay(failCount int, sleep time.Duration, backoffExponentBase float64) time.Duration {
```

Comment:
After the recent changes, I think we ended up overloading the responsibilities of this function and made `RunWithExpBackoff` a bit inconsistent, since it will now do something else than its name suggests for negative bases. Considering that, I think we should prevent `RunWithExpBackoff` from accepting negative bases or make it work with negative exponent bases (a bit weird but consistent), or at least mention that different behaviour in the comment for `RunWithExpBackoff`.

One more idea that comes to my mind is to use a private
```golang
func runWithBackoff(function func() error, delay func(attempt int), attempts int) (err error) {
	for i := 0; i < attempts; i++ {
		delay(i)
		if err = function(); err == nil {
			break
		}
	}
	return err
}
```
function, which would then be called by the `Run*` ones with different `calculateDelay` implementations. For instance

```golang
func Run(function func() error, attempts int, sleep time.Duration) (err error) {
	return runWithBackoff(function, func(attempt int) {
		time.Sleep(time.Duration(attempt) * sleep)
	}, attempts)
}

func RunWithExpBackoff(function func() error, attempts int, sleep time.Duration, backoffExponentBase float64) (err error) {
	return runWithBackoff(function, func(attempt int) {
		if attempt == 0 {
			return
		}
		expRetry := math.Pow(backoffExponentBase, float64(attempt))
		sleepDuration := time.Duration(expRetry * float64(sleep))
		time.Sleep(sleepDuration)
	}, attempts)
}
```

Reply:
Was thinking the same thing actually. This is getting complicated :P